crates/ruff_linter/src/rules/pylint/rules/unnecessary_lambda.rs

```diff
@@ -0,0 +1,209 @@
+            let call_kwargs: Vec<&Expr> = arguments
+                .keywords
+                .iter()
+                .map(|kw| &kw.value)
+                .collect::<Vec<_>>();
+
+            // Collect all positional arguments (e.g., `lambda x, y: func(x, y)`).
+            let call_posargs: Vec<&Expr> = arguments
+                .args
+                .iter()
+                .filter(|arg| !arg.is_starred_expr())
+                .collect::<Vec<_>>();
+
+            // Ex) `lambda **kwargs: func(**kwargs)`
+            match parameters.kwarg.as_ref() {
+                None => {
+                    if !call_kwargs.is_empty() {
+                        return;
+                    }
+                }
+                Some(kwarg) => {
+                    let [call_kwarg] = &call_kwargs[..] else {
+                        return;
+                    };
+
+                    let Expr::Name(ast::ExprName { id, .. }) = call_kwarg else {
+                        return;
+                    };
+
+                    if id.as_str() != kwarg.name.as_str() {
+                        return;
+                    }
+                }
+            }
+
+            // Ex) `lambda *args: func(*args)`
+            match parameters.vararg.as_ref() {
+                None => {
+                    if !call_varargs.is_empty() {
+                        return;
+                    }
+                }
+                Some(vararg) => {
+                    let [call_vararg] = &call_varargs[..] else {
+                        return;
+                    };
+
+                    let Expr::Name(ast::ExprName { id, .. }) = call_vararg else {
+                        return;
+                    };
+
+                    if id.as_str() != vararg.name.as_str() {
+                        return;
+                    }
+                }
+            }
+
+            // Ex) `lambda x, y: func(x, y)`
+            let lambda_posargs: Vec<&Parameter> = parameters
+                .args
+                .iter()
+                .map(|ParameterWithDefault { parameter, .. }| parameter)
+                .collect::<Vec<_>>();
+            if call_posargs.len() != lambda_posargs.len() {
+                return;
+            }
+            for (param, arg) in lambda_posargs.iter().zip(call_posargs.iter()) {
+                let Expr::Name(ast::ExprName { id, .. }) = arg else {
+                    return;
+                };
+                if id.as_str() != param.name.as_str() {
+                    return;
+                }
+            }
+        }
+    }
+
+    // The lambda is necessary if it uses one of its parameters _as_ the function call.
+    // Ex) `lambda x, y: x(y)`
+    let names = {
+        let mut finder = NameFinder::default();
+        finder.visit_expr(func);
+        finder.names
+    };
+
+    for name in names {
+        if let Some(binding_id) = checker.semantic().resolve_name(name) {
```

Comment:
So the reason that `lookup_symbol` didn't work here is a bit hard to explain, but consider:

```python
_ = lambda x: f(lambda x: x)(x)
```

Here, we're trying to see if `f(lambda x: x)` contains any references to the `x` in `_ = lambda x`. If we use `lookup_symbol`, it will try to resolve the inner `x` relative to the outer scope, that of `_ = lambda x`, so it thinks it's a reference to the outer `x`. Using `resolve_name` ensures that we use the correct binding which was already computed in the binding phase.